src/settings.ml

```diff
@@ -50,13 +51,34 @@ let resolve_workspace_vars setting =
     | _ -> assert false
     (* name will always be captured *)
   in
+  let setting = Interop.Regexp.replace setting ~regexp ~replacer in
+  (* resolve ${pathSeparator} *)
+  let regexp = Js_of_ocaml.Regexp.regexp "\\$\\{pathSeparator\\}" in
+  let replacer ~matched:_ ~captures:_ ~offset:_ ~string:_ =
+    Char.to_string Path.sep
+  in
   Interop.Regexp.replace setting ~regexp ~replacer
 
 let substitute_workspace_vars setting =
+  (* Windows paths are case-insensitive *)
+  let folder_replace_all =
+    match Platform.t with
+    | Win32 -> String.Caseless.substr_replace_all
+    | _ -> String.substr_replace_all
+  in
   List.fold (Workspace.workspaceFolders ()) ~init:setting ~f:(fun acc folder ->
-      String.substr_replace_all acc
-        ~pattern:(workspace_folder_path folder)
-        ~with_:(workspace_folder_var folder))
+      let acc' =
+        folder_replace_all acc
+          ~pattern:(workspace_folder_path folder)
+          ~with_:(workspace_folder_var folder)
+      in
+      if String.equal acc' acc then
+        (* no ${workspaceFolder:folder_name}, so pointless and verbose-y using
+           cross-platform ${pathSeparator} *)
+        acc
+      else
+        String.substr_replace_all acc' ~pattern:(Char.to_string Path.sep)
+          ~with_:"${pathSeparator}")
```

Comment:
Replacing every separator with `${pathSeparator}` makes the path quite messy.
Is there any benefit to this over replacing all separators with the forward slash?

I haven't had a chance to check, but I think Windows still reads a forward-slash as a path separator.
If it doesn't, the `resolve_workspace_vars` function could just replace forward slashes with `Path.sep`.

If the paths come from detected sandboxes, I think it's unlikely for there to be mixed separators.

Reply:
> Replacing every separator with ${pathSeparator} makes the path quite messy.

It is slightly messy, which is why the PR originally only did it for relative paths rooted at `${workspaceFolder}`. But that ${pathSeparator} change is not as critical as the case-insensitivity change, so I reduced this PR to just the case-insensitivity.

> Is there any benefit to this over replacing all separators with the forward slash?
> I think Windows still reads a forward-slash as a path separator

I don't think it was Windows that had the problem, but some parsing code which split path parts by whatever the OS path separator was (I think the node.js `Node.path.dirname`, etc. was the problem, but I'll revisit it after the Windows case-insensitivity issue). Anyway, changing the path parsing code seemed like a huge blast radius, so I didn't go down that route. Your suggestion of replacing in `resolve_workspace_vars` looks safer.